### DISCUSSION TITLE
NumFocus GSoC 2024 idea list

### DISCUSSION THREAD

**Discussion by EwoutH** [mesa/mesa#1997] [Ideas]
_@projectmesa/maintainers, please feel free extend and edit this starting post directly. Let's make this a shared document!_

**Data collection overhaul**
The datacollector could use a major overhaul to be more flexible, efficient and support multiple agent types by default. This would entail designing a new conceptual model for the datacollector, implementing and testing it, and deploying it and gathering feedback from users.

_See_ 
- https://github.com/projectmesa/mesa/issues/348
- https://github.com/projectmesa/mesa/issues/976
- https://github.com/projectmesa/mesa/discussions/1944
- https://github.com/projectmesa/mesa/pull/1145

Past workaround attempts:
- using multiple schedulers: https://github.com/projectmesa/mesa/pull/1481. The problem is it is too complex.
- adding agent by type vars https://github.com/projectmesa/mesa/pull/1701. The problem is it is too complex.
- exclude none values https://github.com/projectmesa/mesa/pull/1702. The problem is it can't be converted to DF.

**Scale Mesa to millions of agents via mesa-frames**
https://github.com/projectmesa/mesa/discussions/1939

Past discussions on performance (needs to be summarized):
- https://github.com/projectmesa/mesa/discussions/1033
- https://github.com/projectmesa/mesa/issues/798
- https://github.com/projectmesa/mesa/issues/882

**Comment by tpike3** [comment]
Apologies' for the delay on this but here is the guidance on submissions for GSOC. We have until Tuesday

- Each project on the Ideas list should include: **a) a project title/description b) more detailed description of the project (2-5 sentences) c) expected outcomes d) skills required/preferred e) possible mentors f) expected size of project (90, 175 or 350 hour – yes, 3 options starting in 2023). And if possible, an easy, medium or hard rating of each project.** This helps the more inexperience folks not get overwhelmed and they can focus on reviewing easy project ideas.

- Projects should take ~90 hour, ~175 hours or ~350 hours for GSoC contributors to complete. We understand some GSoC contributors will take less time (if they have more knowledge of the topic or codebase) while others could spend more time. It is also possible that a given project idea could be 90 hours, 175 hours or 350 hours depending on the scope of the idea. If the GSoC contributor proposes to do a smaller scope of the idea then it might be a 175 hr project versus if they propose doing more work with their proposal it could be a 350 hour project.

- Have multiple ideas. Even if you are a new organization and only want one or two GSoC contributors showing that you have multiple ideas (**a bare minimum of 4 solid ideas**) is vital. And for an organization hoping for 4 or more GSoC contributors you should have at least 8-10 ideas and having them categorized is a nice touch if you have subcategories of topics within your project.

**Reply by wang-boyu** [reply]
Not sure whether we can reuse any of the ideas from last year: https://github.com/projectmesa/mesa/wiki/Projects-for-Google-Summer-of-Code-2023

**Reply by rht** [reply]
Yes, definitely, as long as they haven't been worked on.

**Reply by quaquel** [reply]
What about hybrid ABM DEVS? There is some solid literature from Ziegler on simulation formalisms. There are a few ABM papers that explicitly refer to this. There are also some libraries out there that already allow for hybrid ABM devs.

**Reply by rht** [reply]
There is [SimPy](https://simpy.readthedocs.io/en/latest/contents.html), but I don't know enough about it to judge how it relates with other DEVS libraries. Are there examples of early seminal models with public code (along the line of Schelling, Boltzmann wealh, Epstein civil violence, etc) implemented by the NetLogo equivalent of DEVS? Or does NetLogo already have discrete event examples? There is paper on the hybrid library by [MASON](https://cs.gmu.edu/~sean/papers/annsim23des.pdf).

I assume the main task would be to write a more comprehensive scheduler on top of the current `DiscreteEventScheduler`?

**Reply by quaquel** [reply]
There is fundamental work by [Ziegler 2018](https://www.sciencedirect.com/book/9780128133705/theory-of-modeling-and-simulation). In this, he briefly outlines how e.g., the game of life can be fully expressed in a discrete event formalism. There is similar work by Vangheluwe who shows how all simulations (including ODEs) are fundamentally expressable using [ the DEVS formalism](https://en.wikipedia.org/wiki/DEVS).

Practically, it would involve a bit more than making a more comprehensive scheduler. Rather, it means rebuilding MESA on top of an EventTree. Practically, this might sound bigger than it. Ziegler makes a distinction between a model and a simulator. A simulator is akin to a numerical method for solving a system of differential equations, while a model specifies the system of equations. So, bringing this to MESA involves separating the model from the simulator. The simulator currently is left implicit (i.e., the calling of model.step()). It probably has some additional knock-on consequences as well.

The payoff of doing all this is that the current behavior would still be there at the core. However, time would be managed by the simulator instead of the model. Moreover, we can now do all kinds of hybrid stuff. For example, in Epstein, agents can schedule their end of jail time instead of subtracting 1 each tick. In a transmission model, likewise, you can schedule becoming infected after an exposure and the recovery after becoming infected. This hybrid stuff can result in massive runtime reductions.

**Comment by jackiekazil** [comment]
Looking at last years ideas and what proposed here -- I like the two that @EwoutH proposed. 
@tpike3 are you formally submitting for Mesa? 
Can we submit both?

**Reply by rht** [reply]
I find that the data collection touches the core API and requires more nuanced knowledge about the problems with the current data collector. And so would be better suited as a NumFOCUS SDG project for a developer with more experience. But then again, shouldn't leave the door closed to a student who could bring a fresh perspective to the problem.

**Reply by tpike3** [reply]
The application process is really simple. It is a link to the ideas and then number of mentors. 

Then there is 2 critical pieces: 
1. If the number of mentors change by tomorrow at 6 PM US west coast then I have to cancel and resubmit the application. 
2. The projects ideas list follows the GSOC format, we have a minimum of 4 and then I added the TOC as a "nice touch" 

@jackiekazil  (and everyone else) feel free to edit the wiki until tomorrow, but please just follow the format and add to the TOC.

**Reply by rht** [reply]
Link to the wiki: https://github.com/projectmesa/mesa/wiki/Projects-for-Google-Summer-of-Code-2024.

**Reply by jackiekazil** [reply]
That is a great page. This makes me realize how far we have come, but also how much work still needs to be done.
